src/Bridge/Doctrine/MongoDB/ItemDataProvider.php

```diff
@@ -0,0 +1,111 @@
+use Doctrine\ODM\MongoDB\DocumentRepository;
+use ApiPlatform\Core\Bridge\Doctrine\MongoDB\Extension\QueryItemExtensionInterface;
+
+/**
+ * Item data provider for the Doctrine MongoDB.
+ */
+class ItemDataProvider implements ItemDataProviderInterface
+{
+    private $managerRegistry;
+    private $propertyNameCollectionFactory;
+    private $propertyMetadataFactory;
+    private $itemExtensions;
+    private $decorated;
+
+    /**
+     * @param ManagerRegistry                        $managerRegistry
+     * @param PropertyNameCollectionFactoryInterface $propertyNameCollectionFactory
+     * @param PropertyMetadataFactoryInterface       $propertyMetadataFactory
+     * @param QueryItemExtensionInterface[]          $itemExtensions
+     * @param ItemDataProviderInterface|null         $decorated
+     */
+    public function __construct(ManagerRegistry $managerRegistry, PropertyNameCollectionFactoryInterface $propertyNameCollectionFactory, PropertyMetadataFactoryInterface $propertyMetadataFactory, array $itemExtensions = [], ItemDataProviderInterface $decorated = null)
+    {
+        $this->managerRegistry = $managerRegistry;
+        $this->propertyNameCollectionFactory = $propertyNameCollectionFactory;
+        $this->propertyMetadataFactory = $propertyMetadataFactory;
+        $this->itemExtensions = $itemExtensions;
+        $this->decorated = $decorated;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getItem(string $resourceClass, $id, string $operationName = null, bool $fetchData = false)
+    {
+        if ($this->decorated) {
+            try {
+                return $this->decorated->getItem($resourceClass, $id, $operationName, $fetchData);
+            } catch (ResourceClassNotSupportedException $resourceClassNotSupportedException) {
+                // Ignore it
+            }
+        }
+
+        $manager = $this->managerRegistry->getManagerForClass($resourceClass);
+        if (null === $manager) {
+            throw new ResourceClassNotSupportedException();
+        }
+
+        $identifierValues = explode('-', $id);
+        $identifiers = [];
+        $i = 0;
+
+        foreach ($this->propertyNameCollectionFactory->create($resourceClass) as $propertyName) {
+            $itemMetadata = $this->propertyMetadataFactory->create($resourceClass, $propertyName);
+
+            $identifier = $itemMetadata->isIdentifier();
+            if (null === $identifier || false === $identifier) {
+                continue;
+            }
+
+            if (!isset($identifierValues[$i])) {
+                throw new InvalidArgumentException(sprintf('Invalid identifier "%s".', $id));
+            }
+
+            $identifiers[$propertyName] = $identifierValues[$i];
+            ++$i;
+        }
+
+        if (!$fetchData || $manager instanceof DocumentManager) {
+            // @todo: Check potential problems with using the $id instead of $identifiers
+            return $manager->getReference($resourceClass, $id);
```

Comment:
There are severals options for compound keys, i would however not include support for this in this PR.